File: app/src/main/java/com/jiujiu/mvvmTemplate/ui/base/BaseFragment.java
package com.jiujiu.mvvmTemplate.ui.base;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.LayoutRes;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;

import javax.inject.Inject;

import dagger.android.support.AndroidSupportInjection;
import dagger.android.support.DaggerFragment;

public abstract class BaseFragment<T extends ViewDataBinding, V extends BaseViewModel> extends DaggerFragment {

    @Inject
    ViewModelProvider.Factory factory;

    private T mBinding;
    private V mViewModel;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mViewModel = generateViewmodel();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        mBinding = DataBindingUtil.inflate(inflater, getLayoutId(), container, false);
        mBinding.setLifecycleOwner(this);
        return mBinding.getRoot();
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mBinding.setVariable(getBindingVariableId(), mViewModel);
        mBinding.executePendingBindings();
    }

    protected abstract int getBindingVariableId();

    @LayoutRes
    protected abstract int getLayoutId();

    private V generateViewmodel() {
        return ViewModelProviders.of(this, factory).get(getViewModelType());
    }

    protected abstract Class<V> getViewModelType();


    public T getBinding() {
        return mBinding;
    }

    public V getViewModel() {
        return mViewModel;
    }
}

File: app/src/main/java/com/jiujiu/mvvmTemplate/ui/base/BaseFragment.kt
package com.jiujiu.mvvmTemplate.ui.base

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import dagger.android.support.DaggerFragment
import org.jetbrains.anko.AnkoLogger
import javax.inject.Inject

abstract class BaseFragment<T : ViewDataBinding, V : BaseViewModel> : DaggerFragment(), AnkoLogger {

    @Inject
    lateinit var factory: ViewModelProvider.Factory

    lateinit var binding: T
        private set
    lateinit var viewModel: V
        private set

    protected abstract val bindingVariableId: Int

    @get:LayoutRes
    protected abstract val layoutId: Int

    protected abstract val viewModelType: Class<V>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = generateViewmodel()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = DataBindingUtil.inflate(inflater, layoutId, container, false)
        binding.lifecycleOwner = this
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.setVariable(bindingVariableId, viewModel)
        binding.executePendingBindings()
    }

    private fun generateViewmodel(): V {
        return ViewModelProviders.of(this, factory).get(viewModelType)
    }
}
